develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md

```diff
@@ -0,0 +1,456 @@
+
+        public void LogError(string message)
+        {
+            _protocol.Log(message, LogType.Error, LogLevel.NoLogging);
+        }
+
+        public void LogWarning(string message)
+        {
+            _protocol.Log(message, LogType.Error, LogLevel.NoLogging);
+        }
+
+        public void LogInfo(string message)
+        {
+            _protocol.Log(message, LogType.Information, LogLevel.NoLogging);
+        }
+
+        public void LogDebug(string message)
+        {
+            _protocol.Log(message, LogType.DebugInfo, LogLevel.Level2);
+        }
+
+        public void LogTrace(string message)
+        {
+            _protocol.Log(message, LogType.DebugInfo, LogLevel.Level4);
+        }
+    }
+}
+
+```
+
+#### Connecting with the gNMI service
+
+By calling [Connect](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient.Connect), you make a connection with the endpoint.
+
+```csharp
+var client = new GnmiClient((uint)protocol.DataMinerID, (uint)protocol.ElementID, protocol.ElementName, config, logger);
+client.Connect()
+```
+
+In case you want to get notified when the connection state changes, you can register on the [ConnectionStateChanged](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient.ConnectionStateChanged) event.
+
+```csharp
+client.ConnectionStateChanged += OnConnectionStateChanged
+
+private void OnConnectionStateChanged(object sender, EventArgs e)
+{
+    protocol.Log($"Connection state changed to {client.IsConnected}");
+}
+```
+
+> [!NOTE]
+> In case you are setting up a secure channel, it is important that the server certificate is issued to the hostname configured in the [DataSourceConfiguration](xref:Skyline.DataMiner.Helper.OpenConfig.Models.DataSourceConfiguration). The whole certificate chain needs to be trusted and must not be expired. There is no option to disable this. When this is not the desired behavior, you will need to fall back to insecure HTTP.
+
+### Disconnecting
+
+To disconnect from the endpoint, you can call [Disconnect](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient.Disconnect). As GnmiClient implements [IDisposable](xref:System.IDisposable), you can call [Dispose](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient.Dispose) on the client as well, which will call the [Disconnect](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient.Disconnect) for you.
+
+To disconnect when the element has stopped, you can use the QAction's Dispose.
+
+```csharp
+public class QAction : IDisposable
+{
+   private bool disposed;
+   private GnmiClient gnmiClient;
+
+   public void Dispose()
+   {
+      Dispose(true);
+      GC.SuppressFinalize(this);
+   }
+
+   protected virtual void Dispose(bool disposing)
+   {
+      if (disposing && !disposed)
+      {
+         disposed = true;
+         gnmiClient?.Dispose();
+         gnmiClient = null;
+      }
+   }
+}
+```
+
+> [!NOTE]
+> Making a QAction implement IDisposable is only supported in DataMiner version 10.2.9 and up.
```

Comment:
Is this note really necessary, since we've already said that the minimum version for everything here is 10.3.3?

Reply:
The idea behind it was that if system developers see that this is an option, know that they won't be able to use this just yet in any connector but that there's a minimum required version of 10.2.9.